File: src/hetzner_code/commands/ssh.py
from __future__ import annotations

from pathlib import Path

from hetzner_code import ssh_util, state
from hetzner_code.config import REMOTE_CODE_DIR
from hetzner_code.forwards import normalize_forward
from hetzner_code.run import HcodeError


def _resolve_cwd(
    instance: state.Instance, repo_name: str | None, worktree: str | None
) -> str | None:
    if worktree:
        candidates = [r for r in instance.repos if worktree in r.worktrees]
        if repo_name:
            candidates = [r for r in candidates if r.name == repo_name]
        if not candidates:
            raise HcodeError(
                f"no worktree '{worktree}' found"
                + (f" on repo '{repo_name}'" if repo_name else "")
                + f" (see `hcode status {instance.name}`)"
            )
        if len(candidates) > 1:
            names = ", ".join(r.name for r in candidates)
            raise HcodeError(
                f"worktree '{worktree}' exists on more than one repo ({names}) "
                "— disambiguate with --repo"
            )
        return f"{REMOTE_CODE_DIR}/{candidates[0].name}-{worktree}"

    if repo_name:
        if not any(r.name == repo_name for r in instance.repos):
            available = ", ".join(r.name for r in instance.repos) or "(none)"
            raise HcodeError(
                f"'{repo_name}' isn't on '{instance.name}'. On it: {available}"
            )
        return f"{REMOTE_CODE_DIR}/{repo_name}"

    if len(instance.repos) == 1:
        return f"{REMOTE_CODE_DIR}/{instance.repos[0].name}"
    return None


def ssh(
    *,
    instance_name: str,
    repo_name: str | None,
    worktree: str | None,
    forwards: tuple[str, ...],
) -> int:
    instance = state.load(instance_name)
    cwd = _resolve_cwd(instance, repo_name, worktree)
    normalized = [normalize_forward(f) for f in forwards]
    return ssh_util.interactive(
        instance.ip, Path(instance.login_key_path), cwd=cwd, forwards=normalized
    )

File: internal/commands/ssh.go
package commands

import (
	"fmt"
	"strings"

	"github.com/dikee/hcode/internal/config"
	"github.com/dikee/hcode/internal/forwards"
	"github.com/dikee/hcode/internal/run"
	"github.com/dikee/hcode/internal/sshutil"
	"github.com/dikee/hcode/internal/state"
)

func resolveCwd(instance state.Instance, repoName, worktree string) (string, error) {
	if worktree != "" {
		var candidates []state.Repo
		for _, r := range instance.Repos {
			for _, w := range r.Worktrees {
				if w == worktree {
					candidates = append(candidates, r)
					break
				}
			}
		}
		if repoName != "" {
			filtered := candidates[:0]
			for _, r := range candidates {
				if r.Name == repoName {
					filtered = append(filtered, r)
				}
			}
			candidates = filtered
		}
		if len(candidates) == 0 {
			suffix := ""
			if repoName != "" {
				suffix = fmt.Sprintf(" on repo '%s'", repoName)
			}
			return "", run.Errorf(
				"no worktree '%s' found%s (see `hcode status %s`)", worktree, suffix, instance.Name,
			)
		}
		if len(candidates) > 1 {
			names := make([]string, len(candidates))
			for i, r := range candidates {
				names[i] = r.Name
			}
			return "", run.Errorf(
				"worktree '%s' exists on more than one repo (%s) — disambiguate with --repo",
				worktree, strings.Join(names, ", "),
			)
		}
		return fmt.Sprintf("%s/%s-%s", config.RemoteCodeDir, candidates[0].Name, worktree), nil
	}

	if repoName != "" {
		found := false
		for _, r := range instance.Repos {
			if r.Name == repoName {
				found = true
				break
			}
		}
		if !found {
			return "", run.Errorf("'%s' isn't on '%s'. On it: %s", repoName, instance.Name, reposList(instance))
		}
		return fmt.Sprintf("%s/%s", config.RemoteCodeDir, repoName), nil
	}

	if len(instance.Repos) == 1 {
		return fmt.Sprintf("%s/%s", config.RemoteCodeDir, instance.Repos[0].Name), nil
	}
	return "", nil
}

// SSHOptions are every flag `hcode ssh` accepts.
type SSHOptions struct {
	InstanceName string
	RepoName     string
	Worktree     string
	Forwards     []string
}

// SSH runs `hcode ssh`, returning the exit code of the interactive
// session.
func SSH(opts SSHOptions) (int, error) {
	instance, err := state.Load(opts.InstanceName)
	if err != nil {
		return 1, err
	}
	cwd, err := resolveCwd(instance, opts.RepoName, opts.Worktree)
	if err != nil {
		return 1, err
	}
	normalized := make([]string, 0, len(opts.Forwards))
	for _, f := range opts.Forwards {
		nf, err := forwards.NormalizeForward(f)
		if err != nil {
			return 1, err
		}
		normalized = append(normalized, nf)
	}
	return sshutil.Interactive(instance.IP, instance.LoginKeyPath, cwd, normalized), nil
}
